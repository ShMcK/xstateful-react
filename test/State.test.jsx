// @flow

import React from 'react';
import renderer from 'react-test-renderer';
import { Machine } from 'xstate';
import { createStatefulMachine } from '@avaragado/xstateful';

import { createReactMachine } from '..';

const machine = Machine({
    key: 'test',
    initial: 'first',
    states: {
        first: {},
        second: {},
        third: {},
    },
});

const expectMatch = expectedMatch => props =>
    props.match === expectedMatch ? 'pass' : 'fail';

const pass = () => 'pass';
const fail = () => 'fail';

const fixtures = [
    {
        title: 'is string',
        config: {
            machine,
        },
        render: (Provider, State) => (
            <Provider>
                <State is="first">{expectMatch(true)}</State>
                <State is="second">{expectMatch(false)}</State>
            </Provider>
        ),
    },

    {
        title: 'is array',
        config: {
            machine,
        },
        render: (Provider, State) => (
            <Provider>
                <State is={['first', 'third']}>{expectMatch(true)}</State>
                <State is={['second', 'third']}>{expectMatch(false)}</State>
            </Provider>
        ),
    },

    {
        title: 'is function',
        config: {
            machine,
        },
        render: (Provider, State) => (
            <Provider>
                <State is={value => value === 'first'}>
                    {expectMatch(true)}
                </State>
                <State is={value => value === 'second'}>
                    {expectMatch(false)}
                </State>
            </Provider>
        ),
    },

    {
        title: 'not string',
        config: {
            machine,
        },
        render: (Provider, State) => (
            <Provider>
                <State not="first">{expectMatch(false)}</State>
                <State not="second">{expectMatch(true)}</State>
            </Provider>
        ),
    },

    {
        title: 'not array',
        config: {
            machine,
        },
        render: (Provider, State) => (
            <Provider>
                <State not={['first', 'third']}>{expectMatch(false)}</State>
                <State not={['second', 'third']}>{expectMatch(true)}</State>
            </Provider>
        ),
    },

    {
        title: 'not function',
        config: {
            machine,
        },
        render: (Provider, State) => (
            <Provider>
                <State not={value => value === 'first'}>
                    {expectMatch(false)}
                </State>
                <State not={value => value === 'second'}>
                    {expectMatch(true)}
                </State>
            </Provider>
        ),
    },

    {
        title: 'no test prop at all is equivalent to a match',
        config: {
            machine,
        },
        render: (Provider, State) => (
            <Provider>
                <State>{expectMatch(true)}</State>
            </Provider>
        ),
    },

    {
        title: 'passes component/render/children props down',
        config: {
            machine,
        },
        render: (Provider, State) => (
            <Provider>
                <State is="first">{expectMatch(true)}</State>
                <State is="first" render={pass} />
                <State is="first" component={pass} />
                <State is="second">{expectMatch(false)}</State>
                <State is="second" render={fail} />
                <State is="second" component={fail} />
            </Provider>
        ),
    },

    {
        title: 'is > not',
        config: {
            machine,
        },
        render: (Provider, State) => (
            <Provider>
                <State is="first" not="first">
                    {expectMatch(true)}
                </State>
            </Provider>
        ),
    },
];

describe('State', () => {
    test('exists', () => {
        const xsf = createStatefulMachine({ machine });
        xsf.init();

        const { State } = createReactMachine(xsf);

        expect(State).toBeDefined();
    });

    test('copes before xsf.init()', () => {
        const xsf = createStatefulMachine({ machine });

        const { Provider, State } = createReactMachine(xsf);

        // before init, state is null.
        const component = renderer.create(
            <Provider>
                <State is="first">{expectMatch(false)}</State>
                <State not="first">{expectMatch(true)}</State>
                <State is={['first', 'third']}>{expectMatch(false)}</State>
                <State not={['first', 'third']}>{expectMatch(true)}</State>
                <State is={value => value === 'first'}>
                    {expectMatch(false)}
                </State>
                <State not={value => value === 'first'}>
                    {expectMatch(true)}
                </State>
            </Provider>,
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    fixtures.forEach(fixture => {
        test(fixture.title, () => {
            const xsf = createStatefulMachine(fixture.config);
            xsf.init();

            const { Provider, State } = createReactMachine(xsf);

            const component = renderer.create(fixture.render(Provider, State));

            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
});
