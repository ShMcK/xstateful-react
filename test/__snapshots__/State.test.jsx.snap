// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`State copes before xsf.init() 1`] = `
Array [
  "pass",
  "pass",
  "pass",
  "pass",
  "pass",
  "pass",
]
`;

exports[`State is > not 1`] = `"pass"`;

exports[`State is array 1`] = `
Array [
  "pass",
  "pass",
]
`;

exports[`State is function 1`] = `
Array [
  "pass",
  "pass",
]
`;

exports[`State is string 1`] = `
Array [
  "pass",
  "pass",
]
`;

exports[`State no test prop at all is equivalent to a match 1`] = `"pass"`;

exports[`State not array 1`] = `
Array [
  "pass",
  "pass",
]
`;

exports[`State not function 1`] = `
Array [
  "pass",
  "pass",
]
`;

exports[`State not string 1`] = `
Array [
  "pass",
  "pass",
]
`;

exports[`State passes component/render/children props down 1`] = `
Array [
  "pass",
  "pass",
  "pass",
  "pass",
]
`;
